File: Provenance/Controller/PVGBAControllerViewController.m
//
//  PVGBAControllerViewController.m
//  Provenance
//
//  Created by James Addyman on 21/03/2015.
//  Copyright (c) 2015 James Addyman. All rights reserved.
//

#import "PVGBAControllerViewController.h"
#import <PVGBA/PVGBAEmulatorCore.h>
#import "PVSettingsModel.h"
#import <AudioToolbox/AudioToolbox.h>

@interface PVGBAControllerViewController ()

@end

@implementation PVGBAControllerViewController

- (void)viewDidLayoutSubviews
{
    [super viewDidLayoutSubviews];
    
    for (JSButton *button in [self.buttonGroup subviews])
    {
        if (![button isMemberOfClass:[JSButton class]])
        {
            continue;
        }
        
        if ([[[button titleLabel] text] isEqualToString:@"A"])
        {
            [button setTag:PVGBAButtonA];
        }
        else if ([[[button titleLabel] text] isEqualToString:@"B"])
        {
            [button setTag:PVGBAButtonB];
        }
    }
    
    [self.leftShoulderButton setTag:PVGBAButtonL];
    [self.rightShoulderButton setTag:PVGBAButtonR];
    [self.startButton setTag:PVGBAButtonStart];
    [self.selectButton setTag:PVGBAButtonSelect];
}

- (void)dPad:(JSDPad *)dPad didPressDirection:(JSDPadDirection)direction
{
    PVGBAEmulatorCore *gbaCore = (PVGBAEmulatorCore *)self.emulatorCore;
    
    [gbaCore releaseGBAButton:PVGBAButtonUp forPlayer:0];
    [gbaCore releaseGBAButton:PVGBAButtonDown forPlayer:0];
    [gbaCore releaseGBAButton:PVGBAButtonLeft forPlayer:0];
    [gbaCore releaseGBAButton:PVGBAButtonRight forPlayer:0];
    
    switch (direction)
    {
        case JSDPadDirectionUpLeft:
            [gbaCore pushGBAButton:PVGBAButtonUp forPlayer:0];
            [gbaCore pushGBAButton:PVGBAButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionUp:
            [gbaCore pushGBAButton:PVGBAButtonUp forPlayer:0];
            break;
        case JSDPadDirectionUpRight:
            [gbaCore pushGBAButton:PVGBAButtonUp forPlayer:0];
            [gbaCore pushGBAButton:PVGBAButtonRight forPlayer:0];
            break;
        case JSDPadDirectionLeft:
            [gbaCore pushGBAButton:PVGBAButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionRight:
            [gbaCore pushGBAButton:PVGBAButtonRight forPlayer:0];
            break;
        case JSDPadDirectionDownLeft:
            [gbaCore pushGBAButton:PVGBAButtonDown forPlayer:0];
            [gbaCore pushGBAButton:PVGBAButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionDown:
            [gbaCore pushGBAButton:PVGBAButtonDown forPlayer:0];
            break;
        case JSDPadDirectionDownRight:
            [gbaCore pushGBAButton:PVGBAButtonDown forPlayer:0];
            [gbaCore pushGBAButton:PVGBAButtonRight forPlayer:0];
            break;
        default:
            break;
    }
    
	[super dPad:dPad didPressDirection:direction];
}

- (void)dPadDidReleaseDirection:(JSDPad *)dPad
{
    PVGBAEmulatorCore *gbaCore = (PVGBAEmulatorCore *)self.emulatorCore;
    
    [gbaCore releaseGBAButton:PVGBAButtonUp forPlayer:0];
    [gbaCore releaseGBAButton:PVGBAButtonDown forPlayer:0];
    [gbaCore releaseGBAButton:PVGBAButtonLeft forPlayer:0];
    [gbaCore releaseGBAButton:PVGBAButtonRight forPlayer:0];
	[super dPadDidReleaseDirection:dPad];
}

- (void)buttonPressed:(JSButton *)button
{
    PVGBAEmulatorCore *gbaCore = (PVGBAEmulatorCore *)self.emulatorCore;
    [gbaCore pushGBAButton:[button tag] forPlayer:0];
    
	[super buttonPressed:button];
}

- (void)buttonReleased:(JSButton *)button
{
    PVGBAEmulatorCore *gbaCore = (PVGBAEmulatorCore *)self.emulatorCore;
    [gbaCore releaseGBAButton:[button tag] forPlayer:0];
	[super buttonReleased:button];
}

- (void)pressStartForPlayer:(NSUInteger)player
{
    PVGBAEmulatorCore *gbaCore = (PVGBAEmulatorCore *)self.emulatorCore;
    [gbaCore pushGBAButton:PVGBAButtonStart forPlayer:player];
	[super pressStartForPlayer:player];
}

- (void)releaseStartForPlayer:(NSUInteger)player
{
    PVGBAEmulatorCore *gbaCore = (PVGBAEmulatorCore *)self.emulatorCore;
    [gbaCore releaseGBAButton:PVGBAButtonStart forPlayer:player];
	[super releaseStartForPlayer:player];
}

- (void)pressSelectForPlayer:(NSUInteger)player
{
    PVGBAEmulatorCore *gbaCore = (PVGBAEmulatorCore *)self.emulatorCore;
    [gbaCore pushGBAButton:PVGBAButtonSelect forPlayer:player];
	[super pressSelectForPlayer:player];
}

- (void)releaseSelectForPlayer:(NSUInteger)player
{
    PVGBAEmulatorCore *gbaCore = (PVGBAEmulatorCore *)self.emulatorCore;
    [gbaCore releaseGBAButton:PVGBAButtonSelect forPlayer:player];
	[super releaseSelectForPlayer:player];
}

@end

File: Provenance/Controller/PVGBAControllerViewController.swift
//  Converted to Swift 4 by Swiftify v4.1.6640 - https://objectivec2swift.com/
//
//  PVGBAControllerViewController.swift
//  Provenance
//
//  Created by James Addyman on 21/03/2015.
//  Copyright (c) 2015 James Addyman. All rights reserved.
//

import PVSupport

fileprivate extension JSButton {
    var buttonTag: PVGBAButton {
        get {
            return PVGBAButton(rawValue: tag)!
        }
        set {
            tag = newValue.rawValue
        }
    }
}

class PVGBAControllerViewController: PVControllerViewController<PVGBASystemResponderClient> {

    override func layoutViews() {
        buttonGroup?.subviews.forEach {
            guard let button = $0 as? JSButton else {
                return
            }

            if (button.titleLabel?.text == "A") {
                button.buttonTag =  .a
            } else if (button.titleLabel?.text == "B") {
                button.buttonTag =  .b
            }
        }

        leftShoulderButton?.buttonTag =  .l
        rightShoulderButton?.buttonTag =  .r

        startButton?.buttonTag =  .start
        selectButton?.buttonTag =  .select
    }

    override func dPad(_ dPad: JSDPad, didPress direction: JSDPadDirection) {
        emulatorCore.didRelease( .up, forPlayer: 0)
        emulatorCore.didRelease( .down, forPlayer: 0)
        emulatorCore.didRelease( .left, forPlayer: 0)
        emulatorCore.didRelease( .right, forPlayer: 0)
        switch direction {
        case .upLeft:
            emulatorCore.didPush( .up, forPlayer: 0)
            emulatorCore.didPush( .left, forPlayer: 0)
        case .up:
            emulatorCore.didPush( .up, forPlayer: 0)
        case .upRight:
            emulatorCore.didPush( .up, forPlayer: 0)
            emulatorCore.didPush( .right, forPlayer: 0)
        case .left:
            emulatorCore.didPush( .left, forPlayer: 0)
        case .right:
            emulatorCore.didPush( .right, forPlayer: 0)
        case .downLeft:
            emulatorCore.didPush( .down, forPlayer: 0)
            emulatorCore.didPush( .left, forPlayer: 0)
        case .down:
            emulatorCore.didPush( .down, forPlayer: 0)
        case .downRight:
            emulatorCore.didPush( .down, forPlayer: 0)
            emulatorCore.didPush( .right, forPlayer: 0)
        case .none:
            break
        }
        super.dPad(dPad, didPress: direction)
    }

    override func dPadDidReleaseDirection(_ dPad: JSDPad) {
        emulatorCore.didRelease( .up, forPlayer: 0)
        emulatorCore.didRelease( .down, forPlayer: 0)
        emulatorCore.didRelease( .left, forPlayer: 0)
        emulatorCore.didRelease( .right, forPlayer: 0)
        super.dPadDidReleaseDirection(dPad)
    }

    override func buttonPressed(_ button: JSButton) {
        emulatorCore.didPush(button.buttonTag, forPlayer: 0)
        super.buttonPressed(button)
    }

    override func buttonReleased(_ button: JSButton) {
        emulatorCore.didRelease(button.buttonTag, forPlayer: 0)
        super.buttonReleased(button)
    }

    override func pressStart(forPlayer player: Int) {
        emulatorCore.didPush( .start, forPlayer: player)
        super.pressStart(forPlayer: player)
    }

    override func releaseStart(forPlayer player: Int) {
        emulatorCore.didRelease( .start, forPlayer: player)
        super.releaseStart(forPlayer: player)
    }

    override func pressSelect(forPlayer player: Int) {
        emulatorCore.didPush( .select, forPlayer: player)
        super.pressSelect(forPlayer: player)
    }

    override func releaseSelect(forPlayer player: Int) {
        emulatorCore.didRelease( .select, forPlayer: player)
        super.releaseSelect(forPlayer: player)
    }
}
